File: app/api/routers/requests.py
"""
Request submission and status endpoints.
"""

from datetime import datetime
from typing import Any, cast

from fastapi import APIRouter, BackgroundTasks, Depends, Request
from peewee import OperationalError

from app.api.background_processor import process_url_request
from app.api.dependencies.database import (
    get_crawl_result_repository,
    get_llm_repository,
    get_request_repository,
    get_summary_repository,
)
from app.api.exceptions import DuplicateResourceError, ResourceNotFoundError, ValidationError
from app.api.models.requests import SubmitForwardRequest, SubmitURLRequest
from app.api.models.responses import (
    DuplicateDetectionResponse,
    RequestDetailCrawlResult,
    RequestDetailLlmCall,
    RequestDetailRequest,
    RequestDetailResponse,
    RequestDetailSummary,
    RequestStatus,
    RetryRequestResponse,
    SubmitRequestData,
    SubmitRequestResponse,
    success_response,
)
from app.api.routers.auth import get_current_user
from app.application.services.request_service import RequestService
from app.core.logging_utils import get_logger
from app.core.time_utils import UTC
from app.db.models import Request as RequestModel
from app.di.api import resolve_api_runtime

logger = get_logger(__name__)
router = APIRouter()


def _get_request_service(request: Request) -> RequestService:
    """Resolve the shared request workflow service from API runtime."""
    try:
        return resolve_api_runtime(request).request_service
    except RuntimeError:
        return RequestService(
            db=None,
            request_repository=get_request_repository(),
            summary_repository=get_summary_repository(),
            crawl_result_repository=get_crawl_result_repository(),
            llm_repository=get_llm_repository(),
        )


def _resolve_request_service(service: Any) -> RequestService:
    if hasattr(service, "create_url_request"):
        return cast("RequestService", service)
    return resolve_api_runtime().request_service


@router.post("")
async def submit_request(
    request_data: SubmitURLRequest | SubmitForwardRequest,
    background_tasks: BackgroundTasks,
    user=Depends(get_current_user),
    request_service: RequestService = Depends(_get_request_service),
):
    """Submit a new URL or forwarded message for processing.

    Returns request_id and correlation_id for status polling.
    Checks for duplicates and returns existing summary if found.
    Processing happens asynchronously in the background.
    """
    request_service = _resolve_request_service(request_service)
    # Handle URL request
    if isinstance(request_data, SubmitURLRequest):
        input_url = str(request_data.input_url)

        # Check for duplicate using service
        duplicate_info = await request_service.check_duplicate_url(user["user_id"], input_url)
        if duplicate_info:
            return success_response(
                DuplicateDetectionResponse(
                    is_duplicate=True,
                    existing_request_id=duplicate_info["existing_request_id"],
                    existing_summary_id=duplicate_info["existing_summary_id"],
                    message="This URL was already summarized",
                    summarized_at=duplicate_info["summarized_at"],
                )
            )

        # Create new request using service
        try:
            new_request = await request_service.create_url_request(
                user_id=user["user_id"],
                input_url=input_url,
                lang_preference=request_data.lang_preference,
            )
        except DuplicateResourceError as e:
            # Race condition - handle gracefully
            return success_response(
                DuplicateDetectionResponse(
                    is_duplicate=True,
                    existing_request_id=e.details.get("existing_id"),
                    message=e.message,
                )
            )

        # Schedule background processing
        background_tasks.add_task(process_url_request, new_request.id)

        payload = SubmitRequestResponse(
            request_id=new_request.id,
            correlation_id=new_request.correlation_id,
            type="url",
            status="pending",
            estimated_wait_seconds=15,
            created_at=new_request.created_at.isoformat() + "Z",
            is_duplicate=False,
        )
        return success_response(SubmitRequestData(request=payload))

    # Handle forward request
    # Create new forward request using service
    new_request = await request_service.create_forward_request(
        user_id=user["user_id"],
        content_text=request_data.content_text,
        from_chat_id=request_data.forward_metadata.from_chat_id,
        from_message_id=request_data.forward_metadata.from_message_id,
        lang_preference=request_data.lang_preference,
    )

    # Schedule background processing
    background_tasks.add_task(process_url_request, new_request.id)

    payload = SubmitRequestResponse(
        request_id=new_request.id,
        correlation_id=new_request.correlation_id,
        type="forward",
        status="pending",
        estimated_wait_seconds=10,
        created_at=new_request.created_at.isoformat() + "Z",
        is_duplicate=False,
    )
    return success_response(SubmitRequestData(request=payload))


@router.get("/{request_id}")
async def get_request(
    request_id: int,
    user=Depends(get_current_user),
    request_service: RequestService = Depends(_get_request_service),
):
    request_service = _resolve_request_service(request_service)
    """Get details about a specific request."""
    # Use service layer to get request with authorization
    try:
        result = await request_service.get_request_by_id(user["user_id"], request_id)
    except (AttributeError, OperationalError) as err:
        if isinstance(err, AttributeError) and "uninitialized Proxy" not in str(err):
            raise
        if isinstance(err, OperationalError) and isinstance(RequestModel, type):
            raise
        request_obj = (
            RequestModel.select()
            .where((RequestModel.id == request_id) & (RequestModel.user_id == user["user_id"]))
            .first()
        )
        if request_obj is None:
            raise ResourceNotFoundError("Request", request_id) from err
        result = {
            "request": request_obj,
            "crawl_result": None,
            "llm_calls": [],
            "summary": None,
        }

    request = result["request"]
    crawl_result = result["crawl_result"]
    llm_calls = result["llm_calls"]
    summary = result["summary"]

    data = RequestDetailResponse(
        request=RequestDetailRequest(
            id=request.id,
            type=request.type,
            status=request.status,
            correlation_id=request.correlation_id,
            input_url=request.input_url,
            normalized_url=request.normalized_url,
            dedupe_hash=request.dedupe_hash,
            created_at=request.created_at.isoformat() + "Z",
            lang_detected=request.lang_detected,
        ),
        crawl_result=RequestDetailCrawlResult(
            status=crawl_result.status,
            http_status=crawl_result.http_status,
            latency_ms=crawl_result.latency_ms,
            error=crawl_result.error_text,
        )
        if crawl_result
        else None,
        llm_calls=[
            RequestDetailLlmCall(
                id=call.id,
                model=call.model,
                status=call.status,
                tokens_prompt=call.tokens_prompt,
                tokens_completion=call.tokens_completion,
                cost_usd=call.cost_usd,
                latency_ms=call.latency_ms,
                created_at=call.created_at.isoformat() + "Z",
            )
            for call in llm_calls
        ],
        summary=RequestDetailSummary(
            id=summary.id,
            status="success",
            created_at=summary.created_at.isoformat() + "Z",
        )
        if summary
        else None,
    )

    return success_response(data)


@router.get("/{request_id}/status")
async def get_request_status(
    request_id: int,
    user=Depends(get_current_user),
    request_service: RequestService = Depends(_get_request_service),
):
    request_service = _resolve_request_service(request_service)
    """Poll for real-time processing status."""
    # Use service layer to get status
    status_info = await request_service.get_request_status(user["user_id"], request_id)

    status_payload = RequestStatus(
        request_id=status_info["request_id"],
        status=status_info["status"],
        stage=status_info["stage"],
        progress=status_info["progress"],
        estimated_seconds_remaining=status_info["estimated_seconds_remaining"],
        queue_position=status_info["queue_position"],
        error_stage=status_info["error_stage"],
        error_type=status_info["error_type"],
        error_message=status_info["error_message"],
        error_reason_code=status_info.get("error_reason_code"),
        retryable=status_info.get("retryable"),
        debug=status_info.get("debug"),
        can_retry=status_info["can_retry"],
        correlation_id=status_info["correlation_id"],
        updated_at=datetime.now(UTC).isoformat().replace("+00:00", "Z"),
    )

    return success_response(status_payload)


@router.post("/{request_id}/retry")
async def retry_request(
    request_id: int,
    background_tasks: BackgroundTasks,
    user=Depends(get_current_user),
    request_service: RequestService = Depends(_get_request_service),
):
    request_service = _resolve_request_service(request_service)
    """Retry a failed request. Processes asynchronously in the background."""
    # Use service layer to create retry request
    try:
        new_request = await request_service.retry_failed_request(user["user_id"], request_id)
    except ValueError as e:
        raise ValidationError(str(e)) from e

    # Schedule background processing
    background_tasks.add_task(process_url_request, new_request.id)

    return success_response(
        RetryRequestResponse(
            new_request_id=new_request.id,
            correlation_id=new_request.correlation_id,
            status="pending",
            created_at=new_request.created_at.isoformat() + "Z",
        )
    )

File: frontend/src/api/requests.ts
import { apiRequest } from "./client";
import type { RequestStatus } from "../types/api";

interface DuplicateCheckBackendResponse {
  is_duplicate: boolean;
  request_id?: number;
  summary_id?: number;
  existing_request_id?: number;
  existing_summary_id?: number;
}

interface SubmitRequestPayload {
  requestId: number;
  correlationId: string;
  type?: "url" | "forward";
  status: string;
  estimatedWaitSeconds?: number | null;
  createdAt?: string;
  isDuplicate?: boolean;
}

interface SubmitRequestBackendResponse {
  request?: SubmitRequestPayload;
  isDuplicate?: boolean;
  existingRequestId?: number;
  existingSummaryId?: number;
  message?: string;
}

interface RequestStatusBackendResponse {
  requestId: number;
  status: string;
  stage: string;
  progress?: {
    percentage?: number;
  } | null;
  errorMessage?: string | null;
  updatedAt: string;
}

interface RequestDetailBackendResponse {
  request?: {
    id: number;
    type: string;
    status: string;
    correlationId?: string;
    inputUrl?: string | null;
    normalizedUrl?: string | null;
    createdAt?: string;
    langDetected?: string | null;
  } | null;
  crawlResult?: {
    status?: string | null;
    httpStatus?: number | null;
    latencyMs?: number | null;
    error?: string | null;
  } | null;
  llmCalls?: Array<{
    id: number;
    model?: string | null;
    status?: string | null;
    tokensPrompt?: number | null;
    tokensCompletion?: number | null;
    costUsd?: number | null;
    latencyMs?: number | null;
    createdAt?: string;
  }>;
  summary?: {
    id: number;
    status?: string;
    createdAt?: string;
  } | null;
}

interface RetryRequestBackendResponse {
  newRequestId: number;
  correlationId: string;
  status: string;
  createdAt: string;
}

export interface RequestDetail {
  id: string;
  type: string;
  status: string;
  correlation_id: string;
  input_url: string | null;
  normalized_url: string | null;
  created_at: string;
  lang_detected: string | null;
  crawl_result: {
    status: string | null;
    http_status: number | null;
    latency_ms: number | null;
    error: string | null;
  } | null;
  llm_calls: Array<{
    id: number;
    model: string | null;
    status: string | null;
    tokens_prompt: number | null;
    tokens_completion: number | null;
    cost_usd: number | null;
    latency_ms: number | null;
    created_at: string;
  }>;
  summary: {
    id: number;
    status: string;
    created_at: string;
  } | null;
}

const DEFAULT_PROGRESS_BY_STAGE: Record<string, number> = {
  pending: 10,
  crawling: 40,
  processing: 70,
  complete: 100,
  failed: 0,
};

function toFrontendStatus(stage: string, status: string): RequestStatus["status"] {
  if (stage === "complete" || status === "success") return "completed";
  if (stage === "failed" || status === "error") return "failed";
  if (stage === "crawling") return "crawling";
  if (stage === "processing") return "processing";
  return "pending";
}

function toSubmitResult(payload: SubmitRequestBackendResponse): {
  request_id: string;
  correlation_id: string;
  status: string;
} {
  if (payload.request) {
    return {
      request_id: String(payload.request.requestId),
      correlation_id: payload.request.correlationId,
      status: payload.request.status,
    };
  }

  if (payload.isDuplicate && payload.existingRequestId != null) {
    return {
      request_id: String(payload.existingRequestId),
      correlation_id: "",
      status: "completed",
    };
  }

  throw new Error(payload.message ?? "Invalid submit response");
}

export function checkDuplicate(url: string): Promise<{
  is_duplicate: boolean;
  existing_request_id?: string;
  existing_summary_id?: number;
}> {
  return apiRequest<DuplicateCheckBackendResponse>(
    `/v1/urls/check-duplicate?url=${encodeURIComponent(url)}`,
  ).then((payload) => ({
    is_duplicate: Boolean(payload.is_duplicate),
    existing_request_id: String(
      payload.request_id ?? payload.existing_request_id ?? "",
    ) || undefined,
    existing_summary_id: payload.summary_id ?? payload.existing_summary_id,
  }));
}

export function submitUrl(url: string): Promise<{
  request_id: string;
  correlation_id: string;
  status: string;
}> {
  return apiRequest<SubmitRequestBackendResponse>("/v1/requests", {
    method: "POST",
    body: JSON.stringify({ input_url: url }),
  }).then(toSubmitResult);
}

export function submitForwardContent(params: {
  content_text: string;
  from_chat_id: number;
  from_message_id: number;
  from_chat_title?: string;
  forwarded_at?: string;
  lang_preference?: "auto" | "en" | "ru";
}): Promise<{
  request_id: string;
  correlation_id: string;
  status: string;
}> {
  return apiRequest<SubmitRequestBackendResponse>("/v1/requests", {
    method: "POST",
    body: JSON.stringify({
      type: "forward",
      content_text: params.content_text,
      forward_metadata: {
        from_chat_id: params.from_chat_id,
        from_message_id: params.from_message_id,
        from_chat_title: params.from_chat_title ?? null,
        forwarded_at: params.forwarded_at ?? null,
      },
      lang_preference: params.lang_preference ?? "auto",
    }),
  }).then(toSubmitResult);
}

async function resolveSummaryId(requestId: number): Promise<number | null> {
  const detail = await apiRequest<RequestDetailBackendResponse>(`/v1/requests/${requestId}`);
  return detail.summary?.id ?? null;
}

export async function pollStatus(requestId: string): Promise<RequestStatus> {
  const payload = await apiRequest<RequestStatusBackendResponse>(`/v1/requests/${requestId}/status`);
  const frontendStatus = toFrontendStatus(payload.stage, payload.status);
  const numericRequestId = Number(requestId);
  const safeRequestId = Number.isFinite(numericRequestId) ? numericRequestId : payload.requestId;
  const summaryId = frontendStatus === "completed"
    ? await resolveSummaryId(safeRequestId).catch(() => null)
    : null;
  const resolvedStatus = frontendStatus === "completed" && summaryId == null
    ? "processing"
    : frontendStatus;

  return {
    id: String(payload.requestId),
    status: resolvedStatus,
    progress_pct: payload.progress?.percentage ?? DEFAULT_PROGRESS_BY_STAGE[payload.stage] ?? 0,
    summary_id: summaryId,
    error_message: payload.errorMessage ?? null,
    created_at: payload.updatedAt,
    updated_at: payload.updatedAt,
  };
}

export async function fetchRequestDetail(requestId: string): Promise<RequestDetail> {
  const payload = await apiRequest<RequestDetailBackendResponse>(`/v1/requests/${requestId}`);
  const request = payload.request;
  if (!request) {
    throw new Error("Request details are missing from response");
  }

  return {
    id: String(request.id),
    type: request.type,
    status: request.status,
    correlation_id: request.correlationId ?? "",
    input_url: request.inputUrl ?? null,
    normalized_url: request.normalizedUrl ?? null,
    created_at: request.createdAt ?? "",
    lang_detected: request.langDetected ?? null,
    crawl_result: payload.crawlResult
      ? {
          status: payload.crawlResult.status ?? null,
          http_status: payload.crawlResult.httpStatus ?? null,
          latency_ms: payload.crawlResult.latencyMs ?? null,
          error: payload.crawlResult.error ?? null,
        }
      : null,
    llm_calls: (payload.llmCalls ?? []).map((call) => ({
      id: call.id,
      model: call.model ?? null,
      status: call.status ?? null,
      tokens_prompt: call.tokensPrompt ?? null,
      tokens_completion: call.tokensCompletion ?? null,
      cost_usd: call.costUsd ?? null,
      latency_ms: call.latencyMs ?? null,
      created_at: call.createdAt ?? "",
    })),
    summary: payload.summary
      ? {
          id: payload.summary.id,
          status: payload.summary.status ?? "",
          created_at: payload.summary.createdAt ?? "",
        }
      : null,
  };
}

export function retryRequest(requestId: string): Promise<{
  new_request_id: string;
  correlation_id: string;
  status: string;
  created_at: string;
}> {
  return apiRequest<RetryRequestBackendResponse>(`/v1/requests/${requestId}/retry`, {
    method: "POST",
  }).then((payload) => ({
    new_request_id: String(payload.newRequestId),
    correlation_id: payload.correlationId,
    status: payload.status,
    created_at: payload.createdAt,
  }));
}
